[3.x][Installer] Add upload progress and improve error handling

asika32764 opened the pull request:
Add upload progress so user can see the upload percentage like before.

![p-2017-11-12-002](https://user-images.githubusercontent.com/1639206/32691953-d1e3397a-c74a-11e7-8b30-786bf92181ab.jpg)

Improve error handler so user will get message if PHP raise fatal error without standard JSON format or HTTP error status.

![p-2017-11-12-003](https://user-images.githubusercontent.com/1639206/32691960-eecce3e2-c74a-11e7-95fb-e20d63691421.jpg)

asika32764 committed:
Add installation progress

asika32764 committed:
Improve error handle

asika32764 committed:
Improve error handle

asika32764 committed:
Use percentage

asika32764 committed:
Improve error handle

asika32764 committed:
Use CSRF Header

asika32764 committed:
Remove unnecessary var

asika32764 committed:
Unexcepted tabs

asika32764 committed:
Use tabls

asika32764 committed:
all use Tabs

asika32764 committed:
Handle unexpected JSON with string.

asika32764 committed:
Progress bar transition

brianteeman commented on plugins/installer/packageinstaller/tmpl/default.php:
```diff
@@ -130,10 +155,30 @@
+				actions.show();
+				progress.hide();
+
+				var message = 'Unknown error.';
```
shouldnt this  be translatable?

asika32764 replied:
Hmm, this is only a fallback default message, should never display. (If this message display, it means an extension is on developing and developer try to make it error with empty or non-JSON output...) 

Do you think it is a good idea to translate a message which may only developers can see? If you think so I can translate it.

brianteeman replied:
Personally yes i do

asika32764 replied:
OK let's translate it.

brianteeman commented:
It would be nice if this was accessible  there are lots of accessible progress bar examples online

Fedik commented on plugins/installer/packageinstaller/tmpl/default.php:
```diff
@@ -130,10 +155,30 @@
+					message = res.message;
+				}
+
+				alert(message);
```
maybe would be better `Joomla.renderMessages()`?

asika32764 committed:
Accessible

asika32764 committed:
Use Joomla.renderMessages()

asika32764 commented:
@brianteeman Add accessible.

@Fedik Use Joomla.renderMessages()

asika32764 committed:
Don't allow double upload

asika32764 committed:
tabs

asika32764 committed:
tabs

Quy commented on plugins/installer/packageinstaller/tmpl/default.php:
```diff
@@ -61,6 +66,10 @@
+			if (uploading) {
+				return;
+			}
+			
```
Remove tabs in line 72.

Quy commented on plugins/installer/packageinstaller/tmpl/default.php:
```diff
@@ -96,6 +105,10 @@
 			e.stopPropagation();
 
 			dragZone.removeClass('hover');
+			
```
Remove tabs in line 108.

Quy commented on plugins/installer/packageinstaller/tmpl/default.php:
```diff
@@ -108,20 +121,44 @@
+				contentType: false,
+				xhr: function () {
+					var xhr = new window.XMLHttpRequest();
+					
```
Remove tabs in line 138.

Quy commented on plugins/installer/packageinstaller/tmpl/default.php:
```diff
@@ -108,20 +121,44 @@
+							percentage.text(number);
+						}
+					}, false);
+					
```
Remove tabs in line 153.

asika32764 committed:
Make all empty line no tabs and spaces

alikon commented:
I have tested this item :white_check_mark: successfully on b82544e892ba9cb0f2262b33c3f6105499003438<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

ghost commented:
@asika32764 how to get "fatal error without standard JSON format or HTTP error status" to test PR?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

asika32764 commented:
Actually I don't know. Maybe return a XML output. This case should never happened.

In this PR, all response with HTTP 200 will go to `done()` method, but the PHP post_max_size error is unable to catch and returned with JSON and HTTP 200 status:

![p-2017-11-12-005](https://user-images.githubusercontent.com/1639206/32700012-b53253ec-c7f9-11e7-8d77-599efb48e991.jpg)

If jQuery try to parse JSON but fail, it will go to `error()` method. And if response is HTTP 4xx-5xx status will go `error()` too. I handled the `error.responseText` for any PHP fatal error and `error.responseJSON` for standard JSON output, so it should has no other cases we will get.

asika32764 commented:
@franz-wohlkoenig Sorry, I misunderstand what you asked. The  "fatal error without standard JSON format or HTTP error status" is the `post_max_size` issue.

Set php.ini post_max_size to a small value and upload a big file, you will get error message. Before this PR, the uploader will return no message.

If you upload an extension with fatal error in it's install script, this PR will also show the error message. Before this PR it will lose respond and running forever.

Quy commented:
This works with drag and drop, but not with `Or browse for file`. Is this intentional?

asika32764 commented:
Yes, `Or browse for file` use file input and post form to upload package. While the drag and drop use ajax (because JS cannot insert value to file input).

asika32764 committed:
Add translated error message.

asika32764 committed:
Add mor info

Quy commented:
On a local install, the progress bar moves to 100% very quickly, and then takes about 10 seconds to complete. So this is an upload indicator only and not an installing indicator too?

asika32764 commented:
Yes, it is only uploading indicator. Joomla didn't implement the installing indicator interface now.

Quy commented:
In Microsoft Edge 41.16299.15.0 on Windows 10, you don't see the progress bar until the very end. Also, should the % be next to the number without the space?

![edge-upload](https://user-images.githubusercontent.com/368084/32985162-1fd82dde-cc69-11e7-84ec-1943f0aab064.gif)

brianteeman commented:
good spot. yes it would read better without the space between the number and the &

asika32764 committed:
Remove space

asika32764 committed:
Add installing progress

asika32764 committed:
Remove sleep

asika32764 commented:
- Removed space
- Added installing progress after uploaded. See:

![up-pro](https://user-images.githubusercontent.com/1639206/32987426-4655d7fa-cd26-11e7-8767-8709048d035d.gif)

---- 

@Quy About the Edge issue, it is a known bug that MS haven't fixed it. See:

- https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/12224510/
- https://stackoverflow.com/questions/43942396/ajax-file-upload-progress-upload-event-does-not-fire-in-edge-browser

asika32764 committed:
Hide upload actions

Quy commented on plugins/installer/packageinstaller/tmpl/default.php:
```diff
@@ -108,20 +123,47 @@
+							progressBar.css('width', number + '%');
+							progressBar.attr('aria-valuenow', number);
+							percentage.text(number);
+							
```
Remove tabs on line 152.

Quy commented on plugins/installer/packageinstaller/tmpl/default.php:
```diff
@@ -181,23 +283,52 @@
+					<div class="bar bar-success"
+						style="width: 0;"
+						role="progressbar"
+						aria-valuenow="25"
```
Should the value be 0?

asika32764 replied:
The `width` value? The initial width value should be `0`.

asika32764 replied:
Ah, I know, you mean the `aria-valuenow`

Yes it should be 0. Fixed.

asika32764 committed:
MISC update

Quy commented on administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini:
```diff
@@ -5,9 +5,13 @@
 PLG_INSTALLER_PACKAGEINSTALLER_SELECT_FILE="Or browse for file"
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_INSTALL_JOOMLA_EXTENSION="Upload & Install Joomla Extension"
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_PACKAGE_FILE="Upload Package File"
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_AND_INSTALL="Upload & Install"
```
Sorry to nitpick. Please move lines 14-16 below line 11.

asika32764 committed:
Modify language

Quy commented:
I have tested this item :white_check_mark: successfully on a00243c9dba19d6c047c2d24b83b22045d07d739<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

brianteeman commented on administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini:
```diff
@@ -5,9 +5,13 @@
+PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_ERROR_UNKNOWN="Error: Unknown error or invalid JSON output."
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_INSTALL_JOOMLA_EXTENSION="Upload & Install Joomla Extension"
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_PACKAGE_FILE="Upload Package File"
-PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_AND_INSTALL="Upload & Install"
```
For silly reasons it is our practice not to remove unused strings. So please put this string back. You can if you want to add a comment above it
;The next line is deprecated and will be removed in J4

asika32764 replied:
OK added the deprecated info

asika32764 committed:
Update en-GB.plg_installer_packageinstaller.ini

Quy commented on administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini:
```diff
@@ -5,9 +5,16 @@
 
 PLG_INSTALLER_PACKAGEINSTALLER_DRAG_FILE_HERE="Drag and drop file here to upload."
 PLG_INSTALLER_PACKAGEINSTALLER_EXTENSION_PACKAGE_FILE="Extension package file"
+PLG_INSTALLER_PACKAGEINSTALLER_INSTALLING="Installing..."
```
Please add space before `...`

Quy commented on administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini:
```diff
@@ -5,9 +5,16 @@
+PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_ERROR_UNKNOWN="Error: Unknown error or invalid JSON output."
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_INSTALL_JOOMLA_EXTENSION="Upload & Install Joomla Extension"
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_PACKAGE_FILE="Upload Package File"
+PLG_INSTALLER_PACKAGEINSTALLER_UPLOADING="Uploading..."
```
Please add space before `...`

asika32764 committed:
Update en-GB.plg_installer_packageinstaller.ini

asika32764 commented:
ok

twister65 commented:
I have tested this item :red_circle: unsuccessfully on 7b6ed585486c7da22f2b7c0d8930477bd94baff8<br /><br />I do not have an alert popup when the package size exceeds the upload_max_filesize value.
Nothing happening, I have only this message : "No Install Type Found".
This problem is recurring on my host where the upload_max_filesize is set to 2Mo.
Great PR, but it does not work as expected.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

asika32764 commented:
@twister65

The test case with error message is `post_max_size`.

PHP won't catch `upload_max_size` error, it will still run process without any uploaded file, so Joomla cannot find extension package file and return a `No Install Type Found` message.

asika32764 commented:
At here we use message instead of alert popup.
https://github.com/joomla/joomla-cms/pull/18554#issuecomment-343708464

twister65 commented:
@asika32764 , if I set post_max_size to 2 MB, and I upload a file larger than 2 MB. Then the message is "No Install Type Found" and there is no alert popup. This message is not relevant.

asika32764 commented:
> Then the message is "No Install Type Found" and there is no alert popup. 

This is expected behavior,the default error message should show as Joomla message, not popup.

This PR try to catch some PHP error which Joomla cannot catch and show as popup to notice user, without this PR, the uploader will only infinite loop without any response. 

> This message is not relevant.

This message is default Joomla install error message, not belongs to this PR.

twister65 commented:
I have tested this item :white_check_mark: successfully on 7b6ed585486c7da22f2b7c0d8930477bd94baff8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

infograf768 commented on plugins/installer/packageinstaller/tmpl/default.php:
```diff
@@ -181,23 +283,52 @@
+				</p>
+			</div>
 		</div>
 
```
Maybe delete empty line

asika32764 replied:
Deleted

asika32764 committed:
Update default.php

infograf768 commented:
I have tested this item :white_check_mark: successfully on 5a68c03c94069f0189d598d5007ad86504a0fd97<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

infograf768 commented:
As the last change is just cosmetic, let's keep this as RTC.
